### DISCUSSION TITLE
H/a processing in pyroSAR SNAP API

### DISCUSSION THREAD

**Discussion by jmarkloew** [johntruckenbrodt/pyroSAR#128] [General]
Hi John,

I have recently started to develop a Sentinel-1 processing chain with pyroSAR. After a first successful test at processing intensity data, I now want to realize a chain for an H/A-decomposition.  I was able to build the followign graph before I encoutered an error.
```python
from pyroSAR.snap.auxil import parse_recipe, parse_node
import os
import pyroSAR

#input parameter 
##list of path to SLCs
my_files = "S1B_IW_SLC__1SDV_20190516T165151_20190516T165219_016272_01E9FB_5C5A.zip"

info= pyroSAR.identify(my_files)

base=os.path.basename(my_files) 
out_name=os.path.splitext(base)[0] + "_Orb_Cal_Deb_ML_Spk_Ha_TC"/p
out_format= "GeoTIFF-BigTIFF" 

##dem path
my_dem= "SRTM30_Demmin_merge_UTM33N.tif"

##selected Polarisation and decomp-features
my_pol= info.polarizations
my_decomp= ["Entropy", "Anisotropy","Alpha"]
decomp_win_size= 5

##build h/a graph

workflow= parse_recipe("base")

read= workflow["Read"]
read.parameters["file"]= my_files
read.parameters["formatName"]= "SENTINEL-1"

aof=workflow["Apply-Orbit-File"]
aof.parameters["orbitType"]= "Sentinel Precise (Auto Download)"
aof.parameters["polyDegree"]= 3
aof.parameters["continueOnFail"]= False

last_node=aof.id

tpd=parse_node("TOPSAR-Deburst")
tpd.parameters["selectedPolarisations"]= my_pol
workflow.insert_node(tpd, before=last_node)

polMat= parse_node("Polarimetric-Matrices")
polMat.parameters["matrix"]= "C2"

workflow.insert_node(polMat, before=tpd.id)
```
here is the error
![HA_error_jupyternb](https://user-images.githubusercontent.com/77681873/108872517-8941d980-75fa-11eb-8582-6fe09d774e57.PNG)

So far I found out that pyroSAR checks against a list of known nodes, when building a graph. But with my very limited python knowledge I concluded that it might not know the nodes related to decomposition methods. Am I right? 
So I hope you can help me solve this problem and as I am pretty new to python feel free to point out any code redundancies I've produced. 

Best regards,
Johannes

**Comment by johntruckenbrodt** [comment]
Hi Johannes, 
it is working in my case so most likely there is just something mixed up in your configuration. First, please check whether the path to the binary programs of the SNAP installation is part of the `PATH` environment variable. You should then be able to run the following in a command shell:
```
gpt Polarimetric-Matrices -h
```
If this works the pyroSAR `parse_node` function should also work. If it does not, you might need to delete the user configuration files under `C:\Users\username\.pyrosar` on Windows. Then you can run the function again and new user configuration files will be created.  
I know, the configuration management is not as user-friendly as it could be :).
Cheers,
John

**Reply by jmarkloew** [reply]
Thanks for the prompt answer. Rookie mistake for not provide specifics about my set up: 
This runs on Ubuntu 18.04 LTS server with SNAP 8.0 and python 3.8. in a conda environment
Does that change anything?
Also, if I remember correctly, I need the Radarsat-2 toolbox to run the decomposition, right

Thanks for the advise
Johannes

**Reply by johntruckenbrodt** [reply]
Running Ubuntu does not change much in this case except of course the path to the configuration files now being under `/home/username/.pyrosar`.  
The [PolarimetricMatricesOp.java](https://github.com/senbox-org/s1tbx/blob/master/rstb/rstb-op-polarimetric-tools/src/main/java/org/csa/rstb/polarimetric/gpf/PolarimetricMatricesOp.java) is in the Radarsat toolbox, which is included in S1TBX. So I don't think you need to install anything extra. If however you can run `gpt` but not the command above then this might indeed be necessary.
